### PR TITLE
FAN-2232: stringsdict support

### PR DESCRIPTION
**Pull request by stegru**
Provides support for pluralisation rules specified in .stringsdict files.

For something in a .stringsdict file, `NSLocalizedString` will return `GSLocalizedString`, which is a `NSString` that contains the dict. When the formatter gets this, it builds a string format according to the pluralisation rules in the dict.

### CODE REVIEW

File: Source/NSString.m [hunk 1]
```diff
@@ -6765,3 +6775,309 @@ - (void) setString: (NSString*)aString
...
+      return self;
+    }
+
+  static dispatch_once_t      onceToken;
+  static NSRegularExpression *regex = nil;
+
+  regex = [NSRegularExpression
+    regularExpressionWithPattern:@"%#@([^@]+)@"
+			 options:NSRegularExpressionCaseInsensitive
+			   error:nil];
```
Comment (will-velazquez) [inline Source/NSString.m:6824]: I'm guessing the dispatch_once_t is leftover code from caching at some point?
Can we remove it alongside the dispatch header includes? Also remove the static qualifier from the regex

File: Source/NSString.m [hunk 2]
```diff
@@ -6765,3 +6775,309 @@ - (void) setString: (NSString*)aString
...
+	  number = va_arg(args, long);
+	}
+      else if ([type isEqualToString:@"lu"])
+	{
+	  number = va_arg(args, unsigned long);
+	}
+      else if ([type isEqualToString:@"1g"])
+	{
+	  double d = va_arg(args, double);
+	  number = d == 0 ? 0 : d == 1 ? 1 : 10;
```
Comment (will-velazquez) [inline Source/NSString.m:6859]: Can we document what the logic is here? I can't quite figure it out

File: Source/NSString.m [hunk 3]
```diff
@@ -6765,3 +6775,309 @@ - (void) setString: (NSString*)aString
...
+
+- (unichar)characterAtIndex:(NSUInteger)index
+{
+  if (index == 999)
+    {
+      NSString *s =
+	[NSString localizedStringWithFormat:self, (unsigned long) 123,
+					    (unsigned long) 555];
+      return 0;
+    }
```
Comment (will-velazquez) [inline Source/NSString.m:6917]: What?

File: Source/NSString.m [hunk 4]
```diff
@@ -6765,3 +6775,309 @@ - (void) setString: (NSString*)aString
...
+      NSString *varKey = [formatKey substringWithRange:[match rangeAtIndex:1]];
+
+      NSDictionary<NSString *, NSString *> *varDict = _dict[varKey];
+
+      // Get the numeric value of the variable (luckily, only "%ld", "%lu", and
+      // "%1g" is used).
+      unsigned long number;
+      NSString	   *type = varDict[@"NSStringFormatValueTypeKey"];
+
+      if ([type isEqualToString:@"ld"])
```
Comment (will-velazquez) [inline Source/NSString.m:6848]: If we're adding these anyway, can we also add
* `%d` -> `int`
* `%u` -> `unsigned`

while we're at it? While we don't use these right now, somebody else (or us in the future) may end up with them

**Review by will-velazquez — CHANGES_REQUESTED** [review]
lgtm, minus the (what I presume) is leftover debugging code

File: Source/NSString.m (hunk 2 above)
Comment (stegru) [reply Source/NSString.m]:
```
      // Casting to an int would mean 0.3 is "zero" and 1.2 would be "one", so
      // check against just 0 or 1 and let any other value be "other". 
```

File: Source/NSString.m [hunk 5]
```diff
@@ -6765,3 +6773,311 @@ - (void) setString: (NSString*)aString
...
+	}
+      else if ([type isEqualToString:@"u"])
+	{
+	  number = va_arg(args, unsigned int);
+	}
+      else if ([type isEqualToString:@"1g"])
+	{
+	  double d = va_arg(args, double);
+      // Casting to an int would mean 0.3 is "zero" and 1.2 would be "one", so
+      // check against just 0 or 1 and let any other value be "other".
```
Comment (will-velazquez) [inline Source/NSString.m:6865]: Missing tab for indent

**Review by will-velazquez — APPROVED** [review]

**Review by brooke-tilley — APPROVED** [review]
looks good and builds 👍

**Pull request merged**